app/pdfstats.py

```diff
@@ -61,15 +61,19 @@ def minutes_to_hours(minutes: Union[float, int]) -> str:
         return f"{minutes} minute{'s' if minutes > 1 else ''}"
 
 
-def highlight_text(
-    sentence: str, fragments: List[str], highlight_class: str = "highlight"
-) -> str:
-    highlighted = sentence
-    for frag in fragments:
-        highlighted = re.sub(
-            re.escape(frag), re.escape(f'<span class="{ highlight_class }">{frag}</span>'), sentence
-        )
-    return highlighted
+def highlight_text(text: str, ranges: List[Tuple[int, int]], class_name="highlight"):
+    output = []
+    prev_end = 0
+
+    for start, end in sorted(ranges):
```

Comment:
By accident, I ran the version on main first, then switched to this branch, but I got a crash here, since you've changed ranges to a list of tuples instead of strings. I've mentioned before that [we need to refresh cached stats when something throws an exception](https://github.com/SuffolkLITLab/FormFyxer/pull/102#discussion_r1166038386), and it looks like that would happen here for everything that's been cached since the last launch, especially because we use the hash of the file contents, so people can't re-upload their files to fix the issue.

Reply:
Good idea but it doesn't seem tied to this PR at all? For now I've just been clearing the /tmp directory when I deploy. I don't see a good reason to keep around the old and invalid stats to save processing on a small and random subset of forms and we're the only ones deploying this.

Reply:
Sorry, I thought we had a volume on the saved PDFs, which was incorrect. It doesn't break as badly as I thought it would.

There is still the issue that existing URLs (which are stable and can be shared) will disappear without this. But we'd need the saved PDFs to hang around for that to work, so I agree, probably too big for this PR.

I'd still like `view/bad-hash-here` to redirect to the home page, so people who have shared URLs are have the open as tabs don't just get 404s.